---
title: "Creating a microbiome analysis data object"
author: "Stephen Johnson"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Create object}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

Analysis in `stromboli` revolves around the manipulation of data objects. This object typically links together the information in the following files:

* Metadata (tab-delimited text file)
* BIOM file
    * Either version 1 or 2 BIOM files should work
    * Regardless of version, the BIOM file should have taxonomy annotations for all OTUs/ASVs prior to importing
* Tree file (newick format)

We can create a data object using the biom, tree, and metadata files included with stromboli in `inst/extdata/`:

```{r setup, include = FALSE}
devtools::load_all("./", TRUE)
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


```{r}

filter.dep <- 2000
rarefy.dep <- 2000

test_metadata <- system.file("extdata", "mapping.txt", package = "stromboli")
test_metadata
test_biom <- system.file("extdata", "constipation.biom", package = "stromboli")
test_biom
test_tree <- system.file("extdata", "constipation.tree", package = "stromboli")
test_tree

data.obj <- load_data(test_biom, test_metadata, tree.file=test_tree)

ind <- rownames(data.obj$meta.dat)[colSums(data.obj$otu.tab) >= filter.dep]
    
cat(rownames(data.obj$meta.dat)[colSums(data.obj$otu.tab) < filter.dep], 'excluded due to insufficient reads!\n')
data.obj <- subset_data(data.obj, ind)
dist.obj <- construct_distance(data.obj)
    
data.obj.rff <- load_data(otu.file, map.file, tree.file, rff=TRUE, dep=rarefy.dep)
dist.obj.rff <- construct_distance(data.obj.rff)

str(data.obj)
```


